src/compiler/checker.ts

```diff
@@ -13960,6 +13960,11 @@ namespace ts {
                 if (type && type !== unknownType) {
                     return getFlowTypeOfReference(node, type);
                 }
+                if (isSourceFile(container)) {
+                    // look up in the source file's locals or exports
+                    const parent = getSymbolOfNode(container);
+                    return createAnonymousType(parent, container.commonJsModuleIndicator ? parent.exports : globals, emptyArray, emptyArray, createIndexInfo(anyType, /*isReadonly*/ false), undefined);

```

Comment:
`createAnonymousType` looks expensive if we call it on every `this` expression not in a class/function. Don't we already get a type for the source file when doing `import a = require("source-file");`?
Also, we should maintain a type for globals; that would fix #14052. But until then `createAnonymousType(globals)` will be really expensive if there are 1000 globals, and there might be!

Reply:
I am also not sure i understand the need for the indexer.. why not just return a "global" type?

Reply:
@mhegazy Do you want an error on any unknown toplevel this access in Javascript? I don't think we should be stricter than Typescript here.

@andy-ms I'll take a look and see what I can do. A global type seems like a good idea.

I don't think the current code is a real performance problem, though, for a couple of reasons:
1. It only happens a couple of times per project, even in large code bases like chrome-devtools-frontend. Compare to `checkObjectLiteral`, which creates a fresh anonymous type for every object literal in a code base.
2. createAnonymousType just saves a pointer to the `globals` symbol table (which does in fact have hundreds of entries just from JS built-ins). The property symbol themselves cache the resolved type if it's requested.